src/ConsoleInputBuilder.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Console;
+
+use Tempest\Console\Exceptions\UnresolvedArgumentsException;
+
+final class ConsoleInputBuilder
+{
+
+    public function __construct(
+        /** @var ConsoleArgumentDefinition[] */
+        protected array $argumentDefinitions,
+        protected ConsoleArgumentBag $argumentBag,
+    ) {
+
+    }
+
+    /**
+     * @return array<ConsoleInputArgument>
+     * @throws UnresolvedArgumentsException
+     */
+    public function build(): array
+    {
+        $passedArguments = $this->argumentBag->all();
+        $definitionArguments = $this->argumentDefinitions;
+        $validArguments = [];
+
+        if (! $definitionArguments && $passedArguments) {
+            throw UnresolvedArgumentsException::fromArguments($passedArguments);
+        }
+
+        foreach ($definitionArguments as $definitionIdx => $definitionArgument) {
+            foreach ($passedArguments as $idx => $argument) {
+                if ($this->matchesDefinition($argument, $definitionArgument)) {
+                    $validArguments[] = $argument;
+
+                    /**
+                     * In case of an input that uses both named and positional arguments,
+                     * we'll remove the named argument from the passed arguments array.
+                     *
+                     * This is to prevent the named argument from being resolved twice.
+                     */
+                    unset($passedArguments[$idx]);
+                    unset($definitionArguments[$definitionIdx]);
+                    continue 2;
```

Comment:
I'd prefer to not having to rely on `continue 2`. Maybe it's an indication that this nested loop should be refactored to a separate (private) function for simplicity?

Reply:
Fixed - had to use reference which I don't love - if you have better idea lemme know

Reply:
Couldn't we determine whether we should `unset` the array key based on the output of `resolveArgument`?

If it's `null`, we won't unset?